### PR TITLE
8338100: C2: assert(!n_loop->is_member(get_loop(lca))) failed: control must not be back in the loop

### PR DESCRIPTION
**Pull request by rwestrel**
The crash occurs because a `Store` is sunk out of a loop that's an
inner loop of an infinite loop. The infinite loop was just found to be
infinite in the current round of loop opts. When that happens the
infinite loop is not properly attached to the rest of the loop tree. As
a consequence, the `IdealLoopTree` instance for the infinite loop and
its children are only partially initialized (`_nest` is not set) and
the structure is an inconsistent state.

When the `Store` is sunk it's reported as belonging to a loop but the
`IdealLoopTree` for that loop is only half populated. As a consequence
a call to `is_dominator` for that loop hits an inconsistency, returns
an incorrect result and the assert fires.

A possible fix would be a point fix that skips that optimization for a
loop that's part of an infinite loop nest. But given basic methods of
loop opts can't be trusted to work in the infinite loop nest, I
suppose similar issues can surface elsewhere.

It's not the first time, we have issues with an infinite loop that's
not properly attached to the loop tree the first time it is
encountered (a NeverBranch is then added and on the next loop passes,
the infinite loop is properly attached to the loop tree). For instance
on a loop opts round, C2 can see that it has no loops and on the next
that it has some.

I propose fixing this by properly attaching the infinite loop to the
loop tree when it's first discovered. A comment in the code seems to
hint that it requires going over the graph again after the
`NeverBranch` is added but I don't think that's case.

I changed the assert in `loopnode.cpp` because it was there to work
around the inconsistency I mentioned above (no loop in a round, some
loops on the next one).

The change in `parse1.cpp` fixes an issue I ran into when testing the
fix. The existing logic doesn't properly detect an exception backedge.

I added the test case from 8336478 to this. The problem there is that
an infinite loop contains a long counted loop. The long counted loop
is transformed into a loop nest which is a 2 step process that
requires 2 rounds of loop opts. But c2 finds an infinite loop in the
middle of the process which causes it to see no more loops and to not
attempt another round of loop opts. The assert fires because it finds
a long counted loop nest that's half transformed. The change I propose
here fixes this too. If we go with this fix, I'll close 8336478 as
duplicate of this one.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8338100](https://bugs.openjdk.org/browse/JDK-8338100): C2: assert(!n_loop-&gt;is_member(get_loop(lca))) failed: control must not be back in the loop (**Bug** - P3)

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Contributors
 * Emanuel Peter `<epeter@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20797/head:pull/20797` \
`$ git checkout pull/20797`

Update a local copy of the PR: \
`$ git checkout pull/20797` \
`$ git pull https://git.openjdk.org/jdk.git pull/20797/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20797`

View PR using the GUI difftool: \
`$ git pr show -t 20797`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20797.diff">https://git.openjdk.org/jdk/pull/20797.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20797#issuecomment-2321790063)

### CODE REVIEW

File: src/hotspot/share/opto/loopnode.cpp [hunk 1]
```diff
@@ -5529,10 +5519,27 @@ int PhaseIdealLoop::build_loop_tree_impl( Node *n, int pre_order ) {
...
           Node* halt = new HaltNode(if_f, frame, "never-taken loop exit reached");
           _igvn.register_new_node_with_optimizer(halt);
-          set_loop(halt, l);
+          set_loop(halt, _ltree_root);
           _igvn.add_input_to(C->root(), halt);
         }
         set_loop(C->root(), _ltree_root);
+        // move to outer most loop with same header
+        l = m_loop;
+        for (;;) {
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopnode.cpp:5528]: Might be cleaner:
```suggestion
        while (true) {
```

File: src/hotspot/share/opto/loopnode.cpp [hunk 2]
```diff
@@ -5529,10 +5519,27 @@ int PhaseIdealLoop::build_loop_tree_impl( Node *n, int pre_order ) {
...
+          IdealLoopTree* next = l->_parent;
+          if (next == nullptr || next->_head != m) {
+            break;
+          }
+          l = next;
+        }
+        // properly insert infinite loop in loop tree
+        sort(_ltree_root, l);
+        // fix child link from parent
+        IdealLoopTree *p = l->_parent;
```
Comment (chhagedorn) [inline src/hotspot/share/opto/loopnode.cpp:5538]:
```suggestion
        IdealLoopTree* p = l->_parent;
```

**Review by TobiHartmann — APPROVED** [review]
Thanks for the detailed summary. The fix looks reasonable to me.

**Review by TobiHartmann — APPROVED** [review]
Looks good.

**Review by chhagedorn — APPROVED** [review]
Still good!

**Comment by rwestrel** [comment]
Thanks for the re-reviews @TobiHartmann @chhagedorn

**Pull request closed without merging**